plugins/kubernetes/app/models/watchers/base_watcher.rb

```diff
@@ -8,19 +8,20 @@ class BaseWatcher
     def initialize(watch_stream, log: true)
       @watch_stream, @log, = watch_stream, log
       @start_time = Time.now
+      @enabled = false
     end
 
     def start_watching(&block)
+      @enabled = true
       @watch_stream.each do |notice|
         handle_notice(notice, &block)
+        break if stop_watching?
       end
+      cleanup
     end
 
     def stop_watching
-      if @watch_stream
-        @watch_stream.finish
-        @watch_stream = nil
-      end
+      @enabled = false
     end
```

Comment:
There's one scenario where this doesn't work.  If a separate thread calls `#stop_watching` on this object, the watcher will continue waiting for new data to come down the stream.  Once it gets another notice, it will break from the `#each` loop.  But if it never gets another notice (let's say it was watching a single pod that got deleted), then it will get stuck in that wait loop forever.

I'm not sure the cleanest way to fix this.  We might need to check whether it's the same thread that is calling `#stop_watching`, or a different thread, and handle it in different ways.